File: web_old/app/model/Experiment.js
/*
 * (C) Copyright 2014 The Genomics of Gene Expression Lab, CIPF 
 * (http://bioinfo.cipf.es/aconesawp) and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 3 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * Contributors:
 *     Rafael Hernandez de Diego, rhernandez@cipf.es
 *     Ana Conesa Cegarra, aconesa@cipf.es
 *     
 * THIS FILE CONTAINS THE FOLLOWING COMPONENT DECLARATION
 * - Experiment
 */
Ext.define('SL.model.Experiment', {
    extend: 'Ext.data.Model',
    alias: 'widget.Experiment',
    //Extends the Observable class
    mixins: {Observable: 'SL.view.senchaExtensions.Observable'},
    fields: [
        {name: 'experiment_id'},
        {name: 'title'},
        {name: 'experiment_description'},
        {name: 'public_references'},
        {name: 'is_time_course_type'},
        {name: 'is_case_control_type'},
        {name: 'is_survival_type'},
        {name: 'is_single_condition'},
        {name: 'is_multiple_conditions'},
        {name: 'is_other_type'},
        {name: 'biological_rep_no', defaultValue: 0},
        {name: 'technical_rep_no', defaultValue: 0},
        {name: 'contains_chipseq', defaultValue: 0},
        {name: 'contains_dnaseseq', defaultValue: 0},
        {name: 'contains_methylseq', defaultValue: 0},
        {name: 'contains_mrnaseq', defaultValue: 0},
        {name: 'contains_mirnaseq', defaultValue: 0},
        {name: 'contains_metabolomics', defaultValue: 0},
        {name: 'contains_proteomics', defaultValue: 0},
        {name: 'contains_other', defaultValue: false},
        {name: 'submission_date', convert: function(v, rec) {
                return rec.dateConversion(v);
            }},
        {name: 'last_edition_date', convert: function(v, rec) {
                return rec.dateConversion(v);
            }},
        {name: 'experimentDataDirectory'}
    ],
    experiment_owners: [],
    experiment_members: [],
    /**BC********************************************************************
     **
     **GETTERS AND SETTERS
     **
     *EC********************************************************************/
    getID: function() {
        return this.get('experiment_id');
    },
    setID: function(value) {
        this.set('experiment_id', value);
        this.setChanged();
    },
    getTitle: function() {
        return this.get('title');
    },
    setTitle: function(value) {
        this.set('title', value);
        this.setChanged();
    },
    getDescription: function() {
        return this.get('experiment_description');
    },
    setDescription: function(value) {
        this.set('experiment_description', value);
        this.setChanged();
    },
    getPublicReferences: function() {
        return this.get('public_references');
    },
    setPublicReferences: function(value) {
        this.set('public_references', value);
        this.setChanged();
    },
    isTimeCourseType: function() {
        return this.get('is_time_course_type');
    },
    setIsTimeCourseType: function(value) {
        this.set('is_time_course_type', value);
        this.setChanged();
    },
    isCaseControlType: function() {
        return this.get('is_case_control_type');
    },
    setIsCaseControlType: function(value) {
        this.set('is_case_control_type', value);
        this.setChanged();
    },
    isSurvivalType: function() {
        return this.get('is_survival_type');
    },
    setIsSurvivalType: function(value) {
        this.set('is_survival_type', value);
        this.setChanged();
    },
    isSingleCondition: function() {
        return this.get('is_single_condition');
    },
    setIsSingleCondition: function(value) {
        this.set('is_single_condition', value);
        this.setChanged();
    },
    isMultipleCondition: function() {
        return this.get('is_multiple_conditions');
    },
    setIsMultipleCondition: function(value) {
        this.set('is_multiple_conditions', value);
        this.setChanged();
    },
    isOtherType: function() {
        return this.get('is_other_type');
    },
    setIsOtherType: function(value) {
        this.set('is_other_type', value);
        this.setChanged();
    },
    getContainsChipSeq: function() {
        return this.get('contains_chipseq');
    },
    setContainsChipSeq: function(value) {
        this.set('contains_chipseq', value);
        this.setChanged();
    },
    getContainsDNaseSeq: function() {
        return this.get('contains_dnaseseq');
    },
    setContainsDNaseSeq: function(value) {
        this.set('contains_dnaseseq', value);
        this.setChanged();
    },
    getContainsMethylSeq: function() {
        return this.get('contains_methylseq');
    },
    setContainsMethylSeq: function(value) {
        this.set('contains_methylseq', value);
        this.setChanged();
    },
    getContainsmRNASeq: function() {
        return this.get('contains_mrnaseq');
    },
    setContainsmRNASeq: function(value) {
        this.set('contains_mrnaseq', value);
        this.setChanged();
    },
    getContainsSmallRNASeq: function() {
        return this.get('contains_mirnaseq');
    },
    setContainsSmallRNASeq: function(value) {
        this.set('contains_mirnaseq', value);
        this.setChanged();
    },
    getContainsMetabolomics: function() {
        return this.get('contains_metabolomics');
    },
    setContainsMetabolomics: function(value) {
        this.set('contains_metabolomics', value);
        this.setChanged();
    },
    getContainsProteomics: function() {
        return this.get('contains_proteomics');
    },
    setContainsProteomics: function(value) {
        this.set('contains_proteomics', value);
        this.setChanged();
    },
    getContainsOther: function() {
        return this.get('contains_other');
    },
    setContainsOther: function(value) {
        this.set('contains_other', value);
        this.setChanged();
    },
    getSubmissionDate: function() {
        return this.get("submission_date");
    },
    setSubmissionDate: function(value) {
        this.set("submission_date", value);
        this.setChanged();
    },
    getLastEditionDate: function() {
        return this.get("last_edition_date");
    },
    setLastEditionDate: function(value) {
        this.set("last_edition_date", value);
        this.setChanged();
    },
    getOwners: function() {
        if (this.experiment_owners == null) {
            this.experiment_owners = [];
        }
        return this.experiment_owners;
    },
    setOwners: function(experiment_owners) {
        this.experiment_owners = experiment_owners;
        this.setChanged();
    },
    addOwner: function(newOwner) {
        if (typeof newOwner === "string") {
            var newOwnerID = newOwner;
            newOwner = Ext.create('SL.model.User');
            newOwner.setID(newOwnerID);
        }
        this.getOwners().push(newOwner);
        this.setChanged();
    },
    isOwner: function(anUser) {
        if (typeof anUser === "string") {
            var owners = this.getOwners();
            for (var i in owners) {
                if (owners[i].getID() === anUser) {
                    return true;
                }
            }
            return false;
        } else {
            //TODO: WORKS?
            return this.getOwners().indexOf(anUser) !== -1;
        }
    },
    getMembers: function() {
        if (this.experiment_members == null) {
            this.experiment_members = [];
        }
        return this.experiment_members;
    },
    setMembers: function(experiment_members) {
        this.experiment_members = experiment_members;
        this.setChanged();
    },
    addMember: function(newMember) {
        if (typeof newMember === "string") {
            var newMemberID = newMember;
            newMember = Ext.create('SL.model.User');
            newMember.setID(newMemberID);
        }
        this.getMembers().push(newMember);
        this.setChanged();
    },
    isMember: function(anUser) {
        if (typeof anUser === "string") {
            var members = this.getMembers();
            for (var i in members) {
                if (members[i].getID() === anUser) {
                    return true;
                }
            }
            return false;
        } else {
            //TODO: WORKS?
            return this.getMembers().indexOf(anUser) !== -1;
        }
    },    
    setExperimentDataDirectory: function(value) {
        this.set('experimentDataDirectory', value);
        this.setChanged();
    },
    getExperimentDataDirectory: function() {
        return this.get('experimentDataDirectory');
    },
    /**BC********************************************************************************************
     ***  
     *** OTHER FUNCTIONS  
     ***
     **EC********************************************************************************************/
    getMemento: function() {
        var memento = this.copy();
        memento.setOwners(this.getOwners());
        memento.setMembers(this.getMembers());
        return memento;
    },
    restoreFromMemento: function(memento) {
        if (memento !== null) {
            this.copyFrom(memento);
            this.setOwners(memento.getOwners());
            this.setMembers(memento.getMembers());
            this.setChanged();
        }
    },
    toSimpleJSON: function() {
        var JSON_DATA = this.data;

        var owners_JSON_DATA = [];
        var owners = this.getOwners();
        for (var i in owners) {
            owners_JSON_DATA.push(owners[i].toSimpleJSON());
        }
        //MERGE THE OBJECT
        if (owners_JSON_DATA.length > 0) {
            JSON_DATA["experiment_owners"] = owners_JSON_DATA;
        }

        var members_JSON_DATA = [];
        var members = this.getMembers();
        for (var i in members) {
            members_JSON_DATA.push(members[i].toSimpleJSON());
        }
        //MERGE THE OBJECT
        if (members_JSON_DATA.length > 0) {
            JSON_DATA["experiment_members"] = members_JSON_DATA;
        }

        return JSON_DATA;
    },
    statics: {
        /********************************************************************************      
         * This function creates a new MODEL using the information provided as a JSON
         *  
         * @param jsonData, the data for the model in JSON format
         * @return the model with the loaded data    
         ********************************************************************************/
        loadFromJSON: function(jsonData) {
            var model = Ext.create('SL.model.Experiment');

            var experiment_owners = [];
            if (jsonData['experiment_owners']) {
                var owners_JSON_list = jsonData['experiment_owners'];
                delete jsonData['experiment_owners'];

                for (var i in owners_JSON_list) {
                    experiment_owners.push(Ext.create('SL.model.User', owners_JSON_list[i]));
                }
            }
            model.setOwners(experiment_owners);

            var experiment_members = [];
            if (jsonData['experiment_members']) {
                var members_JSON_list = jsonData['experiment_members'];
                delete jsonData['experiment_members'];

                for (var i in members_JSON_list) {
                    experiment_members.push(Ext.create('SL.model.User', members_JSON_list[i]));
                }
            }
            model.setMembers(experiment_members);
            model.set(jsonData);
            return model;
        }
    },
    dateConversion: function(v) {
        if (v == null || v.length < 8) {
            return "";
        }
        if (v instanceof Date) {
            var formatNumber = function(number, length) {
                var str = '' + number;
                while (str.length < length) {
                    str = '0' + str;
                }
                return str;
            };
            v = v.getFullYear() + "/" + (formatNumber(v.getMonth() + 1, 2)) + "/" + formatNumber(v.getDate(), 2);
            return v;
        }
        if (v.indexOf("/") === -1)
            return v.slice(0, 4) + "/" + v.slice(4, 6) + "/" + v.slice(6, 8);
        else
            return v;
    }
});

File: src/java/classes/Experiment.java
/* ***************************************************************
 *  This file is part of STATegra EMS.
 *
 *  STATegra EMS is free software: you can redistribute it and/or 
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of 
 *  the License, or (at your option) any later version.
 * 
 *  STATegra EMS is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with STATegra EMS.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *  More info http://bioinfo.cipf.es/stategraems
 *  Technical contact stategraemsdev@gmail.com
 *  *************************************************************** */
package classes;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 *
 * @author Rafa Hernández de Diego
 */
public class Experiment {

    String experiment_id;
    String title;
    String experiment_description;
    int biological_rep_no;
    int technical_rep_no;
    int contains_chipseq;
    int contains_dnaseseq;
    int contains_methylseq;
    int contains_mrnaseq;
    int contains_mirnaseq;
    int contains_metabolomics;
    int contains_proteomics;
    boolean contains_other;
    String public_references;
    String submission_date;
    String last_edition_date;
    String[] tags;
    String data_dir_type; //local_directory, ftp_server, seeddms_server...
    String data_dir_host;
    String data_dir_port;
    String data_dir_user;
    String data_dir_pass;
    String data_dir_path;

    User[] experiment_owners;
    User[] experiment_members;

    public Experiment() {
    }

    /**
     * This static function returns a new object using the data contained in the
     * given JSON object (as String).
     *
     * @param jsonString the JSON object
     * @return the new Object.
     */
    public static Experiment fromJSON(JsonElement jsonString) {
        Gson gson = new Gson();
        Experiment experiment = gson.fromJson(jsonString, Experiment.class);
        experiment.adaptDates();
        return experiment;
    }

    public String toJSON() {
        Gson gson = new Gson();
        String jsonString = gson.toJson(this);

        return jsonString;
    }

    //**********************************************************************
    //* GETTERS AND SETTERS ************************************************
    //**********************************************************************
    public String getExperimentID() {
        return experiment_id;
    }

    public void setExperimentID(String experiment_id) {
        this.experiment_id = experiment_id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getExperimentDescription() {
        return experiment_description;
    }

    public void setExperimentDescription(String experiment_description) {
        this.experiment_description = experiment_description;
    }

    public int getBiologicalRepNo() {
        return biological_rep_no;
    }

    public void setBiologicalRepNo(int biological_rep_no) {
        this.biological_rep_no = biological_rep_no;
    }

    public int getTechnicalRepNo() {
        return technical_rep_no;
    }

    public void setTechnicalRepNo(int technical_rep_no) {
        this.technical_rep_no = technical_rep_no;
    }

    public int getContainsChipseq() {
        return contains_chipseq;
    }

    public void setContainsChipseq(int contains_chipseq) {
        this.contains_chipseq = contains_chipseq;
    }

    public int getContainsDNaseseq() {
        return contains_dnaseseq;
    }

    public void setContainsDNaseseq(int contains_dnaseseq) {
        this.contains_dnaseseq = contains_dnaseseq;
    }

    public int getContainsMethylseq() {
        return contains_methylseq;
    }

    public void setContainsMethylseq(int contains_methylseq) {
        this.contains_methylseq = contains_methylseq;
    }

    public int getContainsMRNAseq() {
        return contains_mrnaseq;
    }

    public void setContainsMRNAseq(int contains_mrnaseq) {
        this.contains_mrnaseq = contains_mrnaseq;
    }

    public int getContainsSmallRNAseq() {
        return contains_mirnaseq;
    }

    public void setContainsSmallRNAseq(int contains_mirnaseq) {
        this.contains_mirnaseq = contains_mirnaseq;
    }

    public int getContainsMetabolomics() {
        return contains_metabolomics;
    }

    public void setContainsMetabolomics(int contains_metabolomics) {
        this.contains_metabolomics = contains_metabolomics;
    }

    public int getContainsProteomics() {
        return contains_proteomics;
    }

    public void setContainsProteomics(int contains_proteomics) {
        this.contains_proteomics = contains_proteomics;
    }

    public boolean getContainsOther() {
        return contains_other;
    }

    public void setContainsOther(boolean contains_other) {
        this.contains_other = contains_other;
    }

    public void setContainsAnalysis(String analysisType) {
        if ("ChIP-seq".equals(analysisType)) {
            this.setContainsChipseq(this.getContainsChipseq() + 1);
        } else if ("DNAse-seq".equals(analysisType)) {
            this.setContainsDNaseseq(this.getContainsDNaseseq() + 1);
        } else if ("Methyl-seq".equals(analysisType)) {
            this.setContainsMethylseq(this.getContainsMethylseq() + 1);
        } else if ("mRNA-seq".equals(analysisType)) {
            this.setContainsMRNAseq(this.getContainsMRNAseq() + 1);
        } else if ("smallRNA-seq".equals(analysisType)) {
            this.setContainsSmallRNAseq(this.getContainsSmallRNAseq() + 1);
        } else if ("Metabolomics".equals(analysisType)) {
            this.setContainsMetabolomics(this.getContainsMetabolomics() + 1);
        } else if ("Proteomics".equals(analysisType)) {
            this.setContainsProteomics(this.getContainsProteomics() + 1);
        }
    }

    public String getPublicReferences() {
        return public_references;
    }

    public void setPublicReferences(String public_references) {
        this.public_references = public_references;
    }

    public String getSubmissionDate() {
        return submission_date;
    }

    public void setSubmissionDate(String submission_date) {
        this.submission_date = submission_date;
    }

    public String getLastEditionDate() {
        return last_edition_date;
    }

    public void setLastEditionDate(String last_edition_date) {
        this.last_edition_date = last_edition_date;
    }

    public void adaptDates() {
        if (this.submission_date.contains("-")) {
            String[] aux = this.submission_date.split("T");
            this.submission_date = aux[0].replaceAll("-", "");
        }
        if (this.last_edition_date.contains("-")) {
            String[] aux = this.last_edition_date.split("T");
            this.last_edition_date = aux[0].replaceAll("-", "");
        }

    }

    public User[] getExperimentOwners() {
        return experiment_owners;
    }

    public void setExperimentOwners(User[] experiment_owners) {
        this.experiment_owners = experiment_owners;
    }

    public boolean isOwner(String user_id) {
        for (User experiment_owner : experiment_owners) {
            if (experiment_owner.getUserID().equals(user_id)) {
                return true;
            }
        }
        return false;
    }

    public User[] getExperimentMembers() {
        return experiment_members;
    }

    public void setExperimentMembers(User[] experiment_members) {
        this.experiment_members = experiment_members;
    }

    public boolean isMember(String user_id) {
        for (User experiment_member : experiment_members) {
            if (experiment_member.getUserID().equals(user_id)) {
                return true;
            }
        }
        return false;
    }

    public String getDataDirectoryType() {
        return data_dir_type;
    }

    public void setDataDirectoryType(String data_dir_type) {
        this.data_dir_type = data_dir_type;
    }

    public String getDataDirectoryHost() {
        return data_dir_host;
    }

    public void setDataDirectoryHost(String data_dir_host) {
        this.data_dir_host = data_dir_host;
    }

    public String getDataDirectoryPort() {
        return data_dir_port;
    }

    public void setDataDirectoryPort(String data_dir_port) {
        this.data_dir_port = data_dir_port;
    }

    public String getDataDirectoryUser() {
        return data_dir_user;
    }

    public void setDataDirectoryUser(String data_dir_user) {
        this.data_dir_user = data_dir_user;
    }

    public String getDataDirectoryPass() {
        return data_dir_pass;
    }

    public void setDataDirectoryPass(String data_dir_pass) {
        this.data_dir_pass = data_dir_pass;
    }

    public String getDataDirectoryPath() {
        if (data_dir_path != null) {
            return (data_dir_path.lastIndexOf("/") == data_dir_path.length() - 1) ? data_dir_path : data_dir_path + "/";
        }
        return null;
    }

    public void setDataDirectoryPath(String data_dir_path) {
        this.data_dir_path = data_dir_path;
    }

    public String[] getTags() {
        return tags;
    }

    public void setTags(String[] tags) {
        this.tags = tags;
    }

    public void setTags(String tags) {
        if (tags != null) {
            this.tags = tags.split(", ");
        }
    }

    //***********************************************************************
    //* OTHER FUNCTIONS *****************************************************
    //***********************************************************************
    @Override
    public String toString() {
        return this.toJSON();
    }

    public Map<String, String> getDataDirectoryInformation() {
        HashMap<String, String> info = new HashMap<String, String>();
        info.put("type", this.data_dir_type);
        info.put("host", this.data_dir_host);
        info.put("port", this.data_dir_port);
        info.put("user", this.data_dir_user);
        info.put("pass", this.data_dir_pass);
        info.put("root", this.data_dir_path.replaceFirst("/$", ""));
        return info;
    }

    public String export(String tmpDir, String format, String templatesDir) throws Exception {
        String content = "";
        this.data_dir_pass = "*********";
        if ("json".equalsIgnoreCase(format)) {
            content = this.toJSON();
        } else if ("xml".equalsIgnoreCase(format)) {
            content = "<?xml version=\"1.0\"?>\n";
            content += "<experiment id=\"" + this.getExperimentID() + "\">\n";
            JsonObject experiment = new JsonParser().parse(this.toJSON()).getAsJsonObject();

            content += this.generateXMLContent(experiment, 1);
            content += "</experiment>\n";
        } else {
            ArrayList<Pair> elements = this.processElementContent(templatesDir, this.toJSON());

            if ("html".equals(format)) {
                content = this.generateHTMLContent(elements);
            } else {
                throw new Exception(format + " is not a valid format");
            }
        }

        File file = new File(tmpDir + File.separator + this.experiment_id + "." + format);
        PrintWriter writer = new PrintWriter(file);
        writer.println(content);
        writer.close();
        return file.getAbsolutePath();
    }

    private ArrayList<Pair> processElementContent(String templatesDir, String jsonObject) throws FileNotFoundException {
        ArrayList<Pair> elements = new ArrayList<Pair>();
        ArrayList<Pair> fields = processTemplateFile(templatesDir + File.separator + "experiment-form.json");
        JsonObject analysis = new JsonParser().parse(jsonObject).getAsJsonObject();

        elements.add(new Pair("Study details", "", 0, "title"));
        elements.add(new Pair("", "", 1, "section"));
        for (Pair field : fields) {
            elements.add(new Pair(field.value, this.getJsonElementAsString(analysis.get(field.label)), 2, "field"));
        }

        return elements;
    }

    private ArrayList<Pair> processTemplateFile(String template) throws FileNotFoundException {
        ArrayList<Pair> content = new ArrayList<Pair>();
        JsonParser parser = new JsonParser();
        JsonElement jsonElement = parser.parse(new FileReader(template));

        JsonArray sections = jsonElement.getAsJsonObject().get("content").getAsJsonArray();
        JsonArray fields;
        for (JsonElement element : sections) {
            fields = element.getAsJsonObject().get("fields").getAsJsonArray();
            for (JsonElement field : fields) {
                content.add(new Pair(field.getAsJsonObject().get("name").getAsString(), field.getAsJsonObject().get("label").getAsString(), 0, "field"));
            }
        }

        return content;
    }

    private String getJsonElementAsString(JsonElement element) {
        String value = "";
        if (element == null) {
            return "-";
        } else if (element.isJsonArray()) {
            JsonArray array = element.getAsJsonArray();
            for (JsonElement subelement : array) {
                value = this.getJsonElementAsString(subelement);
            }
        } else if (element.isJsonObject()) {
            JsonObject object = element.getAsJsonObject();

            for (Map.Entry<String, JsonElement> entry : object.entrySet()) {
                value += entry.getKey() + ":" + this.getJsonElementAsString(entry.getValue());
            }
        } else if (element.isJsonPrimitive()) {
            value += element.toString().replaceAll("\"", "") + "\n";
        }
        return value;
    }

    private String generateXMLContent(JsonElement jsonCode, int level) {
        String content = "";
        if (jsonCode.isJsonPrimitive()) {
            content += jsonCode.getAsString();
        } else if (jsonCode.isJsonArray()) {
            content += "\n";
            for (JsonElement subelement : jsonCode.getAsJsonArray()) {
                content += this.generateXMLContent(subelement, level + 1);
            }
            content += "\n";
            content += String.join("", Collections.nCopies(level - 1, "\t"));
        } else if (jsonCode.isJsonObject()) {
            for (Map.Entry<String, JsonElement> entry : jsonCode.getAsJsonObject().entrySet()) {
                content += String.join("", Collections.nCopies(level, "\t")) + "<" + entry.getKey() + ">" + this.generateXMLContent(entry.getValue(), level + 1) + "</" + entry.getKey() + ">\n";
            }
        }
        return content;
    }

    private String generateHTMLContent(ArrayList<Pair> elements) {
        String content
                = "<html>"
                + " <head>"
                + "  <title>STATegra EMS report</title>"
                + "  <link rel=\"stylesheet\" href=\"https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css\">"
                + " </head>"
                + " <body style=\"background: #fbfbfb;\">"
                + "  <div style=\"max-width: 1024px; margin: auto; background: #fff; padding:50px 10px;\">"
                + "   <button value=\"Print this page\" onclick=\"window.print()\" class=\"btn-primary hidden-print btn btn-rigth pull-right\"> <span class=\"glyphicon glyphicon-print\" aria-hidden=\"true\"></span> Print this page </button>"
                + "   <table style=\" width: 100%; \"><tbody>";
        int lastLevel = 0;

        for (Pair element : elements) {

            while (lastLevel > element.level) {
                content += "</tbody></table></td></tr>";
                lastLevel--;
            }
            if ("title".equals(element.type)) {
                content += "<tr><td colspan='2'>" + "<h" + (element.level + 1) + ">" + element.label + "</h" + (element.level + 1) + ">" + "</td></tr>";
            } else if ("section".equals(element.type)) {
                content += "<tr><td style=\"width:20px;\"></td><td><table style=\" width: 100%; \" " + (element.level > 0 ? "class='table table-striped table-bordered'" : "") + " ><tbody>";
            } else {
                content += "<tr><td><b>" + element.label + "</b></td><td>" + element.value.replaceAll("\\n", "<br>") + "</td></tr>";
            }
            lastLevel = element.level;
        }
        content += "</tbody></table>";
        content += "</div></body></html>";
        return content;
    }

    private class Pair {

        String name;
        String label;
        String value;
        int level;
        String type;

        public Pair(String name, String label, String value, int level, String type) {
            this.name = name;
            this.label = label;
            this.value = value;
            this.level = level;
            this.type = type;
        }

        public Pair(String label, String value, int level, String type) {
            this.label = label;
            this.value = value;
            this.level = level;
            this.type = type;
        }

    }
}
